src/natcap/invest/urban_nature_access.py

```diff
@@ -0,0 +1,315 @@
+        2. Warp the population density raster to the target spatial reference
+           and pixel size using bilinear interpolation.
+        3. Convert the warped density raster back to population counts.
+
+    Args:
+        source_population_raster_path (string): The source population raster.
+            Pixel values represent the number of people occupying the pixel.
+        target_population_raster_path (string): The path to where the target,
+            warped population raster will live on disk.
+        target_pixel_size (tuple): A tuple of the pixel size for the target
+            raster.  Passed directly to ``pygeoprocessing.warp_raster``.
+        target_bb (tuple): A tuple of the bounding box for the target raster.
+            Passed directly to ``pygeoprocessing.warp_raster``.
+        target_projection_wkt (string): The Well-Known Text of the target
+            spatial reference fro the target raster.  Passed directly to
+            ``pygeoprocessing.warp_raster``.
+        working_dir (string): The path to a directory on disk.  A new directory
+            is created within this directory for the storage of temporary files
+            and then deleted upon successful completion of the function.
+
+    Returns:
+        ``None``
+    """
+    if not os.path.isdir(working_dir):
+        os.makedirs(working_dir)
+    tmp_working_dir = tempfile.mkdtemp(dir=working_dir)
+    population_raster_info = pygeoprocessing.get_raster_info(
+        source_population_raster_path)
+    pixel_area = numpy.multiply(*population_raster_info['pixel_size'])
+    population_nodata = population_raster_info['nodata'][0]
+
+    population_srs = osr.SpatialReference()
+    population_srs.ImportFromWkt(population_raster_info['projection_wkt'])
+
+    # Convert population pixel area to square km
+    population_pixel_area = (
+        pixel_area * population_srs.GetLinearUnits()) / 1e6
+
+    def _convert_population_to_density(population):
+        """Convert population counts to population per square km.
+
+        Args:
+            population (numpy.array): A numpy array where pixel values
+                represent the number of people who reside in a pixel.
+
+        Returns:
+            """
+        out_array = numpy.full(
+            population.shape, FLOAT32_NODATA, dtype=numpy.float32)
+
+        valid_mask = slice(None)
+        if population_nodata is not None:
+            valid_mask = ~numpy.isclose(population, population_nodata)
+
+        out_array[valid_mask] = population[valid_mask] / population_pixel_area
+        return out_array
+
+    # Step 1: convert the population raster to population density per sq. km
+    density_raster_path = os.path.join(tmp_working_dir, 'pop_density.tif')
+    pygeoprocessing.raster_calculator(
+        [(source_population_raster_path, 1)],
+        _convert_population_to_density,
+        density_raster_path, gdal.GDT_Float32, FLOAT32_NODATA)
+
+    # Step 2: align to the LULC
```

Comment:
I like the direct note that this is really aligning to the LULC. Should the docstring and/or variable names reflect this intent as well? I guess, is there a reason to keep the variable names ambiguous and something like `lulc_target_projection_wkt`? Curious what you think.

Reply:
The reason I had used the `target_` prefix in most of the parameters was because they were being passed directly to `pygeoprocessing.warp_raster` as parameters of the same name.  But you're right, I do think this function should reflect how it's actually being used, and so I've updated the parameter inputs to reflect that they're coming from the LULC.  If and when it becomes useful to abstract this away into a separate function, we can make appropriately general parameter names at that time.

I do think it makes sense to require the particular information needed to be passed in rather than just passing in the LULC.  We're requesting only what's needed to do the operation, which feels right to me.

Patched in 6fa201f4